### ISSUE TITLE
Validate alignments - Segmentation fault with addAlignmentsToBatch

### ISSUE THREAD

**Issue by gregtatum** [mozilla/firefox-translations-training#450]
In the en-ca training #384 I'm getting a crash in Marian from the alignments.

[Taskcluster Log](https://firefox-ci-tc.services.mozilla.com/tasks/aU9XU5XFSh6-DRtgslQgCw/runs/3/logs/public/logs/live.log)
```
Error: Segmentation fault
Error: Aborted from setErrorHandlers()::<lambda(int, siginfo_t*, void*)> in /builds/worker/fetches/marian-source/src/common/logging.cpp:130

marian::data::CorpusBase::  addAlignmentsToBatch  (std::shared_ptr<marian::data::CorpusBatch>,  std::vector<marian::data::SentenceTuple,std::allocator<marian::data::SentenceTuple>> const&) + 0x438
marian::data::Corpus::  toBatch  (std::vector<marian::data::SentenceTuple,std::allocator<marian::data::SentenceTuple>> const&) + 0x1252
marian::data::BatchGenerator<marian::data::CorpusBase>::  fetchBatches  () + 0x1204
marian::ThreadPool::enqueue<marian::data::BatchGenerator<marian::data::CorpusBase>::fetchBatchesAsync()::{lambda()#1}>(marian::data::BatchGenerator<marian::data::CorpusBase>::fetchBatchesAsync()::{lambda()#1}&&)::{lambda()#1}::  operator()  () const + 0x33
std::_Function_handler<std::unique_ptr<std::__future_base::_Result_base,std::__future_base::_Result_base::_Deleter> (),std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<std::deque<std::shared_ptr<marian::data::CorpusBatch>,std::allocator<std::shared_ptr<marian::data::CorpusBatch>>>>,std::__future_base::_Result_base::_Deleter>,std::__future_base::_Task_state<marian::ThreadPool::enqueue<marian::data::BatchGenerator<marian::data::CorpusBase>::fetchBatchesAsync()::{lambda()#1}>(marian::data::BatchGenerator<marian::data::CorpusBase>::fetchBatchesAsync()::{lambda()#1}&&)::{lambda()#1},std::allocator<int>,std::deque<std::shared_ptr<marian::data::CorpusBatch>,std::allocator<std::shared_ptr<marian::data::CorpusBatch>>> ()>::_M_run()::{lambda()#1},std::deque<std::shared_ptr<marian::data::CorpusBatch>,std::allocator<std::shared_ptr<marian::data::CorpusBatch>>>>>::  _M_invoke  (std::_Any_data const&) + 0x51
std::__future_base::_State_baseV2::  _M_do_set  (std::function<std::unique_ptr<std::__future_base::_Result_base,std::__future_base::_Result_base::_Deleter> ()>*,  bool*) + 0x2d
std::__future_base::_Task_state<marian::ThreadPool::enqueue<marian::data::BatchGenerator<marian::data::CorpusBase>::fetchBatchesAsync()::{lambda()#1}>(marian::data::BatchGenerator<marian::data::CorpusBase>::fetchBatchesAsync()::{lambda()#1}&&)::{lambda()#1},std::allocator<int>,std::deque<std::shared_ptr<marian::data::CorpusBatch>,std::allocator<std::shared_ptr<marian::data::CorpusBatch>>> ()>::  _M_run  () + 0xf0
std::thread::_State_impl<std::thread::_Invoker<std::tuple<marian::ThreadPool::reserve(unsigned long)::{lambda()#1}>>>::  _M_run  () + 0x1a5
```

In this message from Jorg: https://groups.google.com/g/marian-nmt/c/PjA-rQJ3Oio

> To answer my own post: it was indeed a bug in my alignment. I had some links outside of throng of tokens in one language.

So I suspect there is an alignment that is broken somehow in our code. We should validate the alignments. I'll investigate.

**Comment by gregtatum** [comment]
The pruned lexical list appears broken.

```
➤ zhead https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task/YjSziwtbQmK_IwMSuh8Zag/runs/0/artifacts/public%2Fbuild%2Flex.s2t.pruned.zst
▁desacord ▁disagreement 0.76670740
▁disc ▁disagreement 0.04628430
rep ▁disagreement 0.04173990
àncies ▁disagreement 0.02436420
ence ▁disagreement 0.02119450
vin ▁disagreement 0.02115630
sa ▁disagreement 0.02092720
ància ▁disagreement 0.01707020
form ▁disagreement 0.00557550
▁els ▁disagreement 0.00553730
on ▁disagreement 0.00538460
itat ▁disagreement 0.00527000
s ▁disagreement 0.00397160
acord ▁disagreement 0.00362790
▁sobre ▁disagreement 0.00202400
▁Disc ▁disagreement 0.00068740
▁hi ▁disagreement 0.00053460
▁el ▁disagreement 0.00049640
cer ▁disagreement 0.00045830
▁ha ▁disagreement 0.00038190
▁entre ▁disagreement 0.00034370
un ▁disagreement 0.00026730
▁un ▁disagreement 0.00015280
▁coalició ▁disagreement 0.00011460
▁resoldre ▁disagreement 0.00011460
▁forts ▁disagreement 0.00011460
▁jo ▁disagreement 0.00007640
▁polítics ▁disagreement 0.00007640
▁profund ▁disagreement 0.00007640
▁conflictes ▁disagreement 0.00007640
▁política ▁disagreement 0.00003820
▁que ▁disagreement 0.00003820
▁la ▁disagreement 0.00003820
▁va ▁disagreement 0.00003820
▁haver ▁disagreement 0.00003820
▁encara ▁disagreement 0.00003820
▁demòcrates ▁disagreement 0.00003820
▁conflicte ▁disagreement 0.00003820
▁disputa ▁disagreement 0.00003820
ctic ▁disagreement 0.00003820
```

**Comment by gregtatum** [comment]
I've validated the correctness of the generated alignments.

https://firefox-ci-tc.services.mozilla.com/tasks/b-7CDsKNQ_Cn7wf3RcxDXw#artifacts

**Comment by gregtatum** [comment]
Here's another run with augmentations turned off for students.

https://firefox-ci-tc.services.mozilla.com/tasks/groups/C9rHNIexTxeG0qRGzgyP0A

**Comment by gregtatum** [comment]
This is still happening with OpusTrainer even with no augmentation. I removed it and used Marian directly in my training and worked around this, but we should fix this to support augmentation in teachers.

**Comment by marco-c** [comment]
Was this fixed by #491?

**Comment by eu9ene** [comment]
Yes, the students are training now and we don't see this error after Marian update
